server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java

```diff
@@ -559,27 +578,66 @@ private void balance() {
+                    undesiredAllocationsExcludingShuttingDownNodes += entry.getValue();
+                }
+            }
             DesiredBalanceReconciler.this.unassignedShards.set(unassignedShards);
-            DesiredBalanceReconciler.this.undesiredAllocations.set(undesiredAllocations);
+            DesiredBalanceReconciler.this.undesiredAllocations.set(totalUndesiredAllocations);
+            DesiredBalanceReconciler.this.undesiredAllocationsExcludingShuttingDownNodes.set(
+                undesiredAllocationsExcludingShuttingDownNodes
+            );
             DesiredBalanceReconciler.this.totalAllocations.set(totalAllocations);
 
-            maybeLogUndesiredAllocationsWarning(totalAllocations, undesiredAllocations, routingNodes.size());
+            maybeLogUndesiredAllocationsWarning(
+                totalAllocations,
+                totalUndesiredAllocations,
+                undesiredAllocationsExcludingShuttingDownNodes,
+                routingNodes.size()
+            );
         }
 
-        private void maybeLogUndesiredAllocationsWarning(int allAllocations, int undesiredAllocations, int nodeCount) {
-            // more shards than cluster can relocate with one reroute
-            final boolean nonEmptyRelocationBacklog = undesiredAllocations > 2L * nodeCount;
-            final boolean warningThresholdReached = undesiredAllocations > undesiredAllocationsLogThreshold * allAllocations;
-            if (allAllocations > 0 && nonEmptyRelocationBacklog && warningThresholdReached) {
-                undesiredAllocationLogInterval.maybeExecute(
-                    () -> logger.warn(
-                        "[{}] of assigned shards ({}/{}) are not on their desired nodes, which exceeds the warn threshold of [{}]",
-                        Strings.format1Decimals(100.0 * undesiredAllocations / allAllocations, "%"),
-                        undesiredAllocations,
-                        allAllocations,
-                        Strings.format1Decimals(100.0 * undesiredAllocationsLogThreshold, "%")
-                    )
-                );
+        private void maybeLogUndesiredAllocationsWarning(
+            int totalAllocations,
+            int undesiredAllocations,
+            int undesiredAllocationsExcludingShuttingDownNodes,
+            int nodeCount
+        ) {
+            // shards that cluster can relocate with one reroute
+            final boolean emptyRelocationBacklog = undesiredAllocations <= 2L * nodeCount;
+            if (totalAllocations == 0 || emptyRelocationBacklog) {
+                return;
+            }
+            final long maxUndesiredAllocations = (long) (undesiredAllocationsLogThreshold * totalAllocations);
+            final boolean warningThresholdReached = undesiredAllocations > maxUndesiredAllocations
+                || undesiredAllocationsExcludingShuttingDownNodes > maxUndesiredAllocations;
+            if (warningThresholdReached) {
```

Comment:
we could also log once if there are no shutdown markers, since it means these two cases are the same. I thought, this would allow using each log separately for the charts we have. We could also entirely get rid of the existing log, and only log for the non-shutting-down case.

Reply:
I do not think we really need both. 
May be you have a use case in mind where this would be helpful?

Reply:
not really. I've just changed the existing metric and log: https://github.com/elastic/elasticsearch/pull/113683/commits/01de59f55a34845488346245664ec65f6842c8ce